### PR TITLE
Add additional checks for seed candidate masking to deal with HCAL geometry complications

### PR DESCRIPTION
**Pull request by hatakeyamak**
Add additional checks for seed candidate masking to deal with HCAL geometry complications (phi segmentation and depth segmentation change versus eta).

#### PR description:

This PR will fix some logical flaw in channel masking used for PF clustering seed finder. When we look for a clustering seed, for HCAL, we usually look at 4 neighbors along swiss-cross, and if the seed-candidate has higher energy than the 4 neighbors, the seed-candidate becomes seed, and neighbors get masked from becoming seed candidates. However, in HCAL, this logic causes complications (1) when the depth segmentation changes from one ieta to another ieta, (2) when the phi segmentation changes from one ieta to another ieta. In such boundary area, even if a channel A is one of four neighbors of channel B, the channel B may not be one of four neighbors of channel A.

Now, this channel masking will be done, in the case of HB/HE/HF, only when the neighbors are strictly on the same depth layer and has the same eta or phi. Then, the seed finder channel masking will happen only when two channels are neighbors for each other.

This change will make the seed finder of this CPU code and the seed finder of the GPU code (under development) consistent and address issues reported on page-15 of:
https://indico.cern.ch/event/986788/contributions/4155121/attachments/2165561/3654817/PFClusterGPUV4.pdf

#### PR validation:

The usual PF validation plots from **Mark Saunders**: @zeratul87 without and with this PR.
http://hep.baylor.edu/msaunder/PFval/HCAL_validation_plots
There are some differences in the endcap, but don't look significant.

In addition, 
deltaSumSeed
shows that, with this PR, we can get a perfect agreement between CPU and GPU seed finders (thx M. Saunders)

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

This is not backport.

@bendavid @jsalfeld @zeratul87

### CODE REVIEW

File: RecoParticleFlow/PFClusterProducer/plugins/LocalMaximumSeedFinder.h [hunk 1]
```diff
@@ -1,6 +1,7 @@
 #ifndef __LocalMaximumSeedFinder_H__
 #define __LocalMaximumSeedFinder_H__
 
+#include "DataFormats/Math/interface/deltaPhi.h"
```
Comment (slava77) [inline RecoParticleFlow/PFClusterProducer/plugins/LocalMaximumSeedFinder.h:4]: please include  in .cc, where it's used

File: RecoParticleFlow/PFClusterProducer/plugins/LocalMaximumSeedFinder.cc [hunk 2]
```diff
@@ -140,8 +140,30 @@ void LocalMaximumSeedFinder::findSeeds(const edm::Handle<reco::PFRecHitCollectio
...
+          case PFLayer::HCAL_ENDCAP:
+          case PFLayer::HF_EM:   // with the current HF setting, we won't see this case
+                                 // but this can come in if we change _nNeighbours for HF.
+          case PFLayer::HF_HAD:  // same as above
+            // HO has only one depth and eta-phi segmentation is regular, so no need to make this check
+            auto const& nei = (*input)[neighbour];
+            if (maybeseed.depth() != nei.depth())
+              continue;  // masking is done only if the neighbor is on the same depth layer as the seed
+            if ((abs(deltaPhi(maybeseed.positionREP().phi(), nei.positionREP().phi())) > dphicut) &&
+                (abs(maybeseed.positionREP().eta() - nei.positionREP().eta()) > detacut))
```
Comment (slava77) [inline RecoParticleFlow/PFClusterProducer/plugins/LocalMaximumSeedFinder.cc:159]:
```suggestion
            if (std::abs(deltaPhi(maybeseed.positionREP().phi(), nei.positionREP().phi())) > dphicut &&
                std::abs(maybeseed.positionREP().eta() - nei.positionREP().eta()) > detacut)
```
is safer, else there is a chance to pick up C int abs.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/32873	78186ff 
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show relatively small changes starting from a small increase in the number of particleFlowClusterHCAL (then propagating downstream); e.g. in Run2018C EGamma wf 136.874
all_OldVSNew_RunEGamma2018Cwf136p874c_recoPFClusters_particleFlowClusterHCAL__reRECO_obj_eta

**Comment by silviodonato** [comment]
+1

**Pull request merged**